packages/lit-html/src/test/directives/ref_test.ts

```diff
@@ -140,41 +140,76 @@ suite('ref', () => {
   });
 
   test('calls callback bound to options.host', () => {
-    let queriedEl: Element | null;
-    const host = {
-      calls: [] as Array<string | undefined>,
+    class Host {
+      bool = false;
+      calls: Array<string | undefined> = [];
+      root = document.createElement('div');
       elCallback(e: Element | undefined) {
         this.calls.push(e?.tagName);
-      },
+      }
+      render() {
+        return render(
+          this.bool
+            ? html`<div ${ref(this.elCallback)}></div>`
+            : html`<span ${ref(this.elCallback)}></span>`,
+          this.root,
+          {host: this}
+        );
+      }
+    }
+
+    const testRef = (
+      host: Host,
+      initialCalls: Array<string | undefined> = []
+    ) => {
+      let queriedEl: Element | null;
+      host.bool = true;
+      host.render();
+      queriedEl = host.root.firstElementChild;
+      assert.equal(queriedEl?.tagName, 'DIV');
+      assert.deepEqual(host.calls, [...initialCalls, 'DIV']);
+
+      host.bool = true;
+      host.render();
+      queriedEl = host.root.firstElementChild;
+      assert.equal(queriedEl?.tagName, 'DIV');
+      assert.deepEqual(host.calls, [...initialCalls, 'DIV']);
+
+      host.bool = false;
+      host.render();
+      queriedEl = host.root.firstElementChild;
+      assert.equal(queriedEl?.tagName, 'SPAN');
+      assert.deepEqual(host.calls, [...initialCalls, 'DIV', undefined, 'SPAN']);
+
+      host.bool = true;
+      host.render();
+      queriedEl = host.root.firstElementChild;
+      assert.equal(queriedEl?.tagName, 'DIV');
+      assert.deepEqual(host.calls, [
+        ...initialCalls,
+        'DIV',
+        undefined,
+        'SPAN',
+        undefined,
+        'DIV',
+      ]);
     };
-    const go = (x: boolean) =>
-      render(
-        x
-          ? html`<div ${ref(host.elCallback)}></div>`
-          : html`<span ${ref(host.elCallback)}></span>`,
-        container,
-        {host}
-      );
 
-    go(true);
-    queriedEl = container.firstElementChild;
-    assert.equal(queriedEl?.tagName, 'DIV');
-    assert.deepEqual(host.calls, ['DIV']);
-
-    go(true);
-    queriedEl = container.firstElementChild;
-    assert.equal(queriedEl?.tagName, 'DIV');
-    assert.deepEqual(host.calls, ['DIV']);
-
-    go(false);
-    queriedEl = container.firstElementChild;
-    assert.equal(queriedEl?.tagName, 'SPAN');
-    assert.deepEqual(host.calls, ['DIV', undefined, 'SPAN']);
-
-    go(true);
-    queriedEl = container.firstElementChild;
-    assert.equal(queriedEl?.tagName, 'DIV');
-    assert.deepEqual(host.calls, ['DIV', undefined, 'SPAN', undefined, 'DIV']);
+    // Test first instance
+    const host1 = new Host();
+    testRef(host1);
+
+    // Test second instance
+    const host2 = new Host();
+    testRef(host2);
```

Comment:
Non actionable - checking understanding. The previous behavior would have added `undefined` to the first testRef assertion for the second host giving `[...initialCalls, undefined, 'DIV']);`, because it would accidentally treat the different Hosts as the same context?

Reply:
Exactly. If we take the fix out, the test would fail on the 2nd host seeing its callback called with `undefined` before it is called with its div, which is what's seen in the original repro: https://codesandbox.io/s/lit-ref-8lwb73?file=/src/index.js